### PR TITLE
Migrate codebase to TypeScript and update configurations

### LINKED ISSUE #4380: codebase lacks type safety
## Description

typescript support is lacking in the codebase

## What is the problem?

mistakes are easy to make

## How can we address it?

add typescript support

## Why is it important?

avoid errors

## Who needs this?

doc maintainers

## When should this happen (use version numbers if needed)?

asap

### PR DESCRIPTION
**Pull request by hichemfantar**
Refactor the codebase by converting JavaScript files to TypeScript, enhancing type safety and updating related configurations accordingly.

closes #4380

### CODE REVIEW

File: website/src/pages/showcase.tsx [hunk 1]
```diff
@@ -18,8 +19,13 @@ const renderApp = (app, i) => <AppBox app={app} key={`app-${app.name}-${i}`} />;
 function Section({
   element = 'section',
   children,
-  className,
   background = 'light',
+  className,
+}: {
+  element?: any;
```
Comment (Abbondanzo) [inline website/src/pages/showcase.tsx:25]: Can this prop be done away with? Looks like no usages in the file and we can default to using `<section>`

File: website/src/pages/showcase.tsx [hunk 2]
```diff
@@ -18,8 +19,13 @@ const renderApp = (app, i) => <AppBox app={app} key={`app-${app.name}-${i}`} />;
...
+  children: React.ReactNode;
+  background?: 'light' | 'dark';
+  className?: string;
```
Comment (Abbondanzo) [inline website/src/pages/showcase.tsx:28]: Same for this one. We don't use it so we can scrap it

File: website/src/theme/BlogSidebar/Desktop/index.d.ts [hunk 3]
```diff
@@ -1,9 +0,0 @@
-/**
- * Copyright (c) Meta Platforms, Inc. and affiliates.
- *
- * This source code is licensed under the MIT license found in the
- * LICENSE file in the root directory of this source tree.
- */
-/// <reference types="react" />
-import type { Props } from "@theme/BlogSidebar/Desktop";
```
Comment (Abbondanzo) [inline website/src/theme/BlogSidebar/Desktop/index.d.ts:8]: For these swizzled components, could you import the `Props` types and use them for each of the exported components?

**Review by Abbondanzo — CHANGES_REQUESTED** [review]
A few changes but overall looks great!

File: website/src/pages/Home/components/ThemeImage/index.tsx [hunk 4]
```diff
@@ -17,6 +17,7 @@ function ThemeImage({lightSrc, darkSrc, className, alt}) {
           mutation.type === 'attributes' &&
           mutation.attributeName === 'data-theme'
         ) {
+          // @ts-expect-error
```
Comment (Simek) [inline website/src/pages/Home/components/ThemeImage/index.tsx:20]: We should avoid the ignores in general, and fix code instead. If there is something that you cannot resolve, let's add a note to the comment describing why it need to be ignored.

File: website/src/theme/MDXComponents/InlineCode.tsx [hunk 5]
```diff
@@ -24,7 +24,9 @@ export default function InlineCode(props) {
 // Currently only supports links, which should be good enough...
 // To enable Markdown support, use the "`md " prefix
 // Syntax example: `md ({nativeEvent: [PressEvent](pressevent)})`
-const MarkdownInlineCode = React.memo(function MarkdownInlineCodeInner(props) {
+const MarkdownInlineCode = React.memo(function MarkdownInlineCodeInner(props: {
+  children: React.ReactNode;
+}) {
```
Comment (Simek) [inline website/src/theme/MDXComponents/InlineCode.tsx:29]:
```suggestion
const MarkdownInlineCode = React.memo(function MarkdownInlineCodeInner(props: PropsWithChildren) {
```

Typing nit.

**Review by Simek — CHANGES_REQUESTED** [review]
Looks good so far! 👍 Left few suggestions above.

File: website/src/pages/showcase.tsx [hunk 6]
```diff
@@ -9,29 +9,20 @@ import React, {useEffect, useState} from 'react';
...
 
 function Section({
-  element = 'section',
   children,
-  className,
   background = 'light',
+}: {
+  children: React.ReactNode;
+  background?: 'light' | 'dark';
 }) {
```
Comment (Simek) [inline website/src/pages/showcase.tsx:24]:
```suggestion
}: PropsWithChildren<{
  background?: 'light' | 'dark';
}>) {
```

Typing nit.

File: website/core/DocsRating.tsx [hunk 7]
```diff
@@ -4,6 +4,11 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
+declare global {
+  interface Window {
+    ga?: (...args: any[]) => void;
+  }
+}
```
Comment (Simek) [inline website/core/DocsRating.tsx:11]:
```suggestion
declare global {
  interface Window {
    ga?: (...args: any[]) => void;
  }
}
```

We should avoid mocking in non-test env and change the `window?.ga` to be optional, it should be sufficient to fix the problem.

Comment (Simek) [reply website/core/DocsRating.tsx:11]: Let's try adding `@types/google.analytics` then, it should handle the missing types:
* https://www.npmjs.com/package/@types/google.analytics

**Comment by Simek** [comment]
I have some `yarn.lock` changes generated when running branch code locally. Can you make sure that lock is up to date?

When `yarn` is run in monorepo root we also run lock deduplication automatically after that, so make sure that you run it from there. There is no need to address it now by hooking it also at website level, since with yarn upgrade it will be redundant anyway.

**Comment by hichemfantar** [comment]
lockfile is now uptodate

**Review by Simek — APPROVED** [review]
Thanks for the quick updates! 🙏 

Tested the changes locally, switching to TypeScript is a nice codebase and DX improvement, LGTM! 👍

**Comment by hichemfantar** [comment]
Awesome let’s get this bad boy merged 🚀

File: website/package.json [hunk 8]
```diff
@@ -14,23 +14,24 @@
...
     "format:examples": "eslint-examples-js --fix && eslint-examples-tsx --fix",
     "prettier": "yarn format:source && yarn format:markdown && yarn format:style",
-    "lint": "eslint ../docs/** blog/** core/** src/**/*.js ./*.js",
+    "lint": "eslint ../docs/** blog/** core/** src/**/*.tsx ./*.ts",
     "lint:examples": "eslint-examples-js && eslint-examples-tsx && tsc-examples",
     "lint:versioned": "eslint versioned_docs/**",
     "lint:markdown": "remark ../docs --quiet -r .remarkrc.mjs",
     "lint:markdown:versioned": "remark ./versioned_docs --quiet -r .remarkrc.mjs",
-    "lint:format": "prettier --check {{core,src}/**/*.js,*.js} ../docs/*.md {versioned_docs/**/*.md,blog/*.md} src/**/*.{scss,css}",
+    "lint:format": "prettier --check {{core,src}/**/*.js,*.js,/**/*.jsx,*.jsx,/**/*.ts,*.ts,/**/*.tsx,*.tsx} ../docs/*.md {versioned_docs/**/*.md,blog/*.md} src/**/*.{scss,css}",
```
Comment (Simek) [inline website/package.json:34]:
```suggestion
    "lint:format": "prettier --check {core,src}/**/*.tsx core/**/*.tsx ../docs/*.md {versioned_docs/**/*.md,blog/*.md} src/**/*.{scss,css}",
```

Same issue as above.

**Review by Simek — APPROVED** [review]

File: website/package.json [hunk 9]
```diff
@@ -49,27 +50,31 @@
     ]
   },
   "dependencies": {
-    "@docusaurus/core": "3.6.0",
-    "@docusaurus/faster": "3.6.0",
-    "@docusaurus/plugin-google-gtag": "3.6.0",
-    "@docusaurus/plugin-pwa": "3.6.0",
-    "@docusaurus/preset-classic": "3.6.0",
+    "@docusaurus/core": "^3.6.3",
```
Comment (Abbondanzo) [inline website/package.json:53]: Could these dependencies be kept pinned?
```suggestion
    "@docusaurus/core": "3.6.3",
```

File: website/package.json [hunk 10]
```diff
@@ -14,23 +14,24 @@
...
     "test": "yarn build",
     "version:cut": "docusaurus docs:version",
-    "format:source": "prettier --write {{core,src}/**/*.js,*.js}",
+    "format:source": "prettier --write '{core,src}/**/*.{js,jsx,ts,tsx}'",
     "format:markdown": "prettier --write ../docs/*.md && prettier --write {versioned_docs/**/*.md,blog/*.md}",
     "format:style": "prettier --write src/**/*.{scss,css}",
     "format:examples": "eslint-examples-js --fix && eslint-examples-tsx --fix",
     "prettier": "yarn format:source && yarn format:markdown && yarn format:style",
-    "lint": "eslint ../docs/** blog/** core/** src/**/*.js ./*.js",
+    "lint": "eslint ../docs/** blog/** core/** src/**/*.tsx ./*.ts",
```
Comment (Abbondanzo) [inline website/package.json:29]: Not that we should be encouraging anyone to write JS, but they could sneak it in at some point:
```suggestion
    "lint": "eslint ../docs/** blog/** '{core,src}/**/*.{js,jsx,ts,tsx}'",
```

File: website/package.json [hunk 11]
```diff
@@ -14,23 +14,24 @@
...
     "lint:examples": "eslint-examples-js && eslint-examples-tsx && tsc-examples",
     "lint:versioned": "eslint versioned_docs/**",
     "lint:markdown": "remark ../docs --quiet -r .remarkrc.mjs",
     "lint:markdown:versioned": "remark ./versioned_docs --quiet -r .remarkrc.mjs",
-    "lint:format": "prettier --check {{core,src}/**/*.js,*.js} ../docs/*.md {versioned_docs/**/*.md,blog/*.md} src/**/*.{scss,css}",
+    "lint:format": "prettier --check {core,src}/**/*.tsx core/**/*.ts ../docs/*.md {versioned_docs/**/*.md,blog/*.md} src/**/*.{scss,css}",
```
Comment (Abbondanzo) [inline website/package.json:34]:
```suggestion
    "lint:format": "prettier --check '{core,src}/**/*.{js,jsx,ts,tsx}' ../docs/*.md {versioned_docs/**/*.md,blog/*.md} src/**/*.{scss,css}",
```

File: website/package.json (hunk 10 above)
Comment (Simek) [reply website/package.json:29]: Prettier errors out, if there are no matching files in declarated paths:
Screenshot 2024-12-03 at 23 03 24

Comment (Abbondanzo) [reply website/package.json:29]: That's if the matching paths are expanded (i.e. within brackets). If you keep it in string format, the pattern doesn't get expanded:
```
yarn run v1.22.22
$ prettier --write '{core,src}/**/*.{js,jsx,ts,tsx}'
core/DocsRating.js 34ms
core/PlatformTag.js 3ms
core/PrismTheme.js 6ms
core/TableRowWithCodeBlock.js 2ms
core/TabsConstants.js 6ms
core/VerticalTable.js 2ms
src/pages/Home/CallToAction/index.js 2ms
src/pages/Home/Community/index.js 5ms
src/pages/Home/components/Section/index.js 2ms
src/pages/Home/components/SectionTitle/index.js 2ms
src/pages/Home/components/ThemeImage/index.js 3ms
src/pages/Home/Framework/index.js 4ms
src/pages/Home/Hero/Devices.js 23ms
src/pages/Home/Hero/FloorBackground.js 1ms
src/pages/Home/Hero/GridBackground.js 2ms
src/pages/Home/Hero/index.js 2ms
src/pages/Home/index.js 1ms
src/pages/Home/Logo.js 1ms
src/pages/Home/Native/index.js 1ms
src/pages/Home/Platforms/FoxFact.js 3ms
src/pages/Home/Platforms/index.js 3ms
src/pages/Home/Watch/index.js 1ms
src/pages/index.js 1ms
src/pages/showcase.js 7ms
src/pages/versions.js 7ms
✨  Done in 0.48s.
```

Comment (Simek) [reply website/package.json:29]: Nice, TIL!

File: plugins/remark-snackplayer/package.json [hunk 12]
```diff
@@ -15,7 +15,7 @@
     "src"
   ],
   "scripts": {
-    "prettier": "prettier --write \"{src/**/*.js,tests/**/*.js,*.md}\"",
+    "prettier": "prettier --write \"{src/**/*.js,src/**/*.jsx,src/**/*.ts,src/**/*.tsx,tests/**/*.js,*.md}\"",
```
Comment (Abbondanzo) [inline plugins/remark-snackplayer/package.json:18]:
```suggestion
    "prettier": "prettier --write '{src,tests}/**/*.{md,js,jsx,ts,tsx}'",
```

File: plugins/remark-lint-no-dead-urls/package.json [hunk 13]
```diff
@@ -14,7 +14,7 @@
     "src/*"
   ],
   "scripts": {
-    "prettier": "prettier --write \"{src/**/*.js,tests/**/*.js,*.md}\"",
+    "prettier": "prettier --write \"{src/**/*.js,src/**/*.jsx,src/**/*.ts,src/**/*.tsx,tests/**/*.js,*.md}\"",
```
Comment (Abbondanzo) [inline plugins/remark-lint-no-dead-urls/package.json:17]:
```suggestion
    "prettier": "prettier --write '{src,tests}/**/*.{md,js,jsx,ts,tsx}'",
```

File: website/package.json (hunk 10 above)
Comment (hichemfantar) [reply website/package.json:29]: was wondering how i can suppress these errors, thanks!

**Review by Abbondanzo — APPROVED** [review]

**Pull request merged**